### DISCUSSION TITLE
Planar output from ns-process-data equirectangular mode has image artifacts, doesn't respect crop parameters

### DISCUSSION THREAD

**Discussion by darrendavid** [nerfstudio-project/nerfstudio#3450] [Q&A]
I logged this [as an issue](https://github.com/nerfstudio-project/nerfstudio/issues/3445) yesterday, but I'm not convinced it's an issue (it may just be me misunderstanding the system) so I'm posting it here as well.

When I pass 8k equirect imagery into `ns-process-data` using [the specified parameters](https://docs.nerf.studio/quickstart/custom_dataset.html#data-equirectangular), the planar output images A) are riddled with white/off-color pixels in dark areas and B) include the bottom 20% region that I specified should be excluded via the `--crop-bottom` param.

My command line is:

`ns-process-data images --camera-type equirectangular --images-per-equirect 14 --crop-bottom 0.2 --data <input_folder> --output-dir <output_folder>`

I've also tried `--crop factor 0 0.2 0 0` instead of `--crop-bottom 0.2` and received the same result.

I have tried converting my source images to PNG, JPG, and TIF - all output still has the same artifacts.

**Expected behavior**
- Planar output images should be free of artifacts
- Planar output images should reflect the 20% bottom crop on the source equirect

**Screenshots**
Below is a source image (7680x3840 360 equirect) and the resulting 14 planar output images from `ns-process-data` that show the pixelated "noise" in the dark areas of the output, as well as the inclusion of the helmet-cam in the output which I specified should be cropped out from the bottom.

**Source Image**

![input_7680x3840_360equirect](https://github.com/user-attachments/assets/cdb7c5bd-6fdd-41f5-bcce-8507e1cbd9f4)

**Planar Output**

![frame_00001](https://github.com/user-attachments/assets/22732de1-bd9a-444f-bc68-9495c84ae4df)
![frame_00002](https://github.com/user-attachments/assets/3858f567-9695-40e3-be76-d256a0427854)
![frame_00003](https://github.com/user-attachments/assets/d68902ef-a0d9-4867-bfa0-7b77f35c98ad)
![frame_00004](https://github.com/user-attachments/assets/6b71c24f-321a-489e-b6b1-25a0c5b4acca)
![frame_00005](https://github.com/user-attachments/assets/44d217ee-6b57-4fb3-930a-9c50ce25dd6f)
![frame_00006](https://github.com/user-attachments/assets/3ed439c6-a5cb-4aae-909f-8c492320a808)
![frame_00007](https://github.com/user-attachments/assets/afd4f97d-be1f-41f7-ad69-2706149307d5)
![frame_00008](https://github.com/user-attachments/assets/54b6fe64-c677-4aa3-9a80-f9652c598da5)
![frame_00009](https://github.com/user-attachments/assets/e458d43f-de2b-4edb-9359-841cc7c80d3d)
![frame_00010](https://github.com/user-attachments/assets/20dee733-d404-4701-b48b-b1fa498bad1f)
![frame_00011](https://github.com/user-attachments/assets/d8c6ca8e-d772-4dd4-8d4a-c143f9fca91c)
![frame_00012](https://github.com/user-attachments/assets/1c9fd02f-1847-4cf9-8ad3-61953f3d4295)
![frame_00013](https://github.com/user-attachments/assets/d75244e7-5aea-46cb-ab2b-bd3dbd8e7f5f)
![frame_00014](https://github.com/user-attachments/assets/45be2726-6693-4b47-8e67-b433b76c2619)

**Additional context**
This may be expected/normal behavior for this system, but I'm new to this workflow and I'm not getting great alignment out of the output, so I'm checking to see if this is all part of the process or if there's something amiss here. I imagine colmap might need the full image (inclusive of the 20% crop on the bottom) to perform alignment? I don't want that in the reconstruction, though.

**Comment by abrahamezzeddine** [comment]
Hello,

I found a potential bug that would cause that issue with equirectangular images. Sometimes, there is overshooting/undershooting and creates values beyond 0-255 at light/dark pixels. This is mostly seen at edges.
The cause is this;


```
def remap_cubic(
    img: torch.Tensor, map_x: torch.Tensor, map_y: torch.Tensor, border_mode: str = "border"
) -> torch.Tensor:
    """Remap image using bicubic interpolation.

    Args:
        img (torch.Tensor): Image tensor
        map_x (torch.Tensor): x mapping
        map_y (torch.Tensor): y mapping
        border_mode (str, optional): What to do with borders. Defaults to "border".

    Returns:
        torch.Tensor: _description_
    """
    batch_size, channels, height, width = img.shape

    grid_x = (map_x / width + 1) * 2 - 1
    grid_y = (map_y / height + 1) * 2 - 1

    if border_mode == "border":
        grid_x = torch.clamp(grid_x, -1, 1)
        grid_y = torch.clamp(grid_y, -1, 1)
    elif border_mode == "wrap":
        grid_x = torch.remainder(grid_x + 1, 2) - 1
        grid_y = torch.remainder(grid_y + 1, 2) - 1

    grid = torch.stack((grid_x, grid_y), dim=-1).unsqueeze(0).expand(batch_size, -1, -1, -1)

    **return torch.nn.functional.grid_sample(img, grid, mode="bicubic", padding_mode="zeros")**
```

Values can become lower than 0 and larger than 255. The values have to be clamped.

Change the function to below;
```
[def remap_cubic(
    img: torch.Tensor, map_x: torch.Tensor, map_y: torch.Tensor, border_mode: str = "border"
) -> torch.Tensor:
    Remap image using bicubic interpolation.

    Args:
        img (torch.Tensor): Image tensor
        map_x (torch.Tensor): x mapping
        map_y (torch.Tensor): y mapping
        border_mode (str, optional): What to do with borders. Defaults to "border".

    Returns:
        torch.Tensor: _description_
    """
    batch_size, channels, height, width = img.shape

    grid_x = (map_x / width + 1) * 2 - 1
    grid_y = (map_y / height + 1) * 2 - 1

    if border_mode == "border":
        grid_x = torch.clamp(grid_x, -1, 1)
        grid_y = torch.clamp(grid_y, -1, 1)
    elif border_mode == "wrap":
        grid_x = torch.remainder(grid_x + 1, 2) - 1
        grid_y = torch.remainder(grid_y + 1, 2) - 1

    grid = torch.stack((grid_x, grid_y), dim=-1).unsqueeze(0).expand(batch_size, -1, -1, -1)

    result = torch.nn.functional.grid_sample(img, grid, mode="bicubic", padding_mode="zeros")
    
    # Clamp the output to the valid range [0, 1] to prevent artifacts
    return torch.clamp(result, 0.0, 1.0)](url)
